### PR TITLE
Update opam-file-format to 2.1.5

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
The opam libraries need a newer version of `opam-file-format` so in this PR I'm updating our vendored copy to the latest release, 2.1.5 and changing it to use the non-deprecated `FullPos` API (which explains why there are so many changes to the dune code)

### CODE REVIEW

File: src/dune_engine/package.ml [hunk 1]
```diff
@@ -99,13 +99,17 @@ module Dependency = struct
...
-    let to_relop : t -> OpamParserTypes.relop = function
-      | Eq -> `Eq
-      | Gte -> `Geq
-      | Lte -> `Leq
-      | Gt -> `Gt
-      | Lt -> `Lt
-      | Neq -> `Neq
+    let to_relop : t -> OpamParserTypes.FullPos.relop =
+      let nopos pelem =
+        OpamParserTypes.FullPos.{ pelem; pos = Opam_file.nopos }
```
Comment (rgrinberg) [inline src/dune_engine/package.ml:104]: No need to open the module. This should be enough:
```
{ OpamParserTypes.FullPos. pelem; pos = Opam_file.nopos }
```

File: src/dune_engine/package.ml [hunk 2]
```diff
@@ -756,21 +761,21 @@ let load_opam_file file name =
...
+    match value.pelem with
+    | String s -> Some [ s ]
+    | List l ->
       let+ l =
-        List.fold_left l ~init:(Some []) ~f:(fun acc v ->
+        List.fold_left l.pelem ~init:(Some []) ~f:(fun acc v ->
             let* acc = acc in
-            match v with
-            | OpamParserTypes.String (_, s) -> Some (s :: acc)
+            match v.OpamParserTypes.FullPos.pelem with
```
Comment (rgrinberg) [inline src/dune_engine/package.ml:777]: We don't qualify module names in fields. So this should be `v.pelem`. If necessary, you can add a type annotation to `v`.

File: src/dune_engine/package.ml [hunk 3]
```diff
@@ -756,21 +761,21 @@ let load_opam_file file name =
...
+            | OpamParserTypes.FullPos.String s -> Some (s :: acc)
```
Comment (rgrinberg) [inline src/dune_engine/package.ml:778]: Same here. Should be `String s`

File: src/dune_rules/install.ml [hunk 4]
```diff
@@ -429,44 +429,37 @@ let gen_install_file entries =
...
 let load_install_file path =
-  let open OpamParserTypes in
+  let open OpamParserTypes.FullPos in
   let file = Io.Untracked.with_lexbuf_from_file path ~f:Opam_file.parse in
-  let fail (fname, line, col) msg =
-    let pos : Lexing.position =
-      { pos_fname = fname; pos_lnum = line; pos_bol = 0; pos_cnum = col }
+  let fail { filename = pos_fname; start; stop } msg =
+    let position_of_loc (pos_lnum, pos_cnum) =
+      Lexing.{ pos_fname; pos_lnum; pos_bol = 0; pos_cnum }
```
Comment (rgrinberg) [inline src/dune_rules/install.ml:437]: Should be `{ Lexing.pos_fname ; .. }`

File: src/dune_engine/package.ml [hunk 5]
```diff
@@ -227,34 +231,35 @@ module Dependency = struct
     <|> let+ name = Name.decode in
         { name; constraint_ = None }
 
-  let rec opam_constraint : Constraint.t -> OpamParserTypes.value =
-    let nopos = Opam_file.nopos in
+  let nopos pelem = OpamParserTypes.FullPos.{ pelem; pos = Opam_file.nopos }
```
Comment (rgrinberg) [inline src/dune_engine/package.ml:234]: `{ OpamParserTypes.FullPos.pelem; pos = Opam_file.nopos }`

Comment (Leonidas-from-XIV) [reply src/dune_engine/package.ml]: Fixed and moved the whole function one up into the module, since I had it defined multiple times.

**Comment by Leonidas-from-XIV** [comment]
This unfortunately doesn't solve #6767 because the latest release, 2.1.5, is older than https://github.com/ocaml/opam-file-format/pull/54. Maybe we can ask the OPAM team to release 2.1.6, as I would prefer to stick to vendoring releases.

**Pull request merged**